### PR TITLE
feat: stop running query

### PR DESCRIPTION
**Pull request by astandrik**
Сloses https://github.com/ydb-platform/ydb-embedded-ui/issues/1062

[Stand](https://nda.ya.ru/t/YRG_fWWK77QsdD)

### Test Status: <span style="color: orange;">⚠️ FLAKY</span>
📊 [Full Report](https://ydb-platform.github.io/ydb-embedded-ui/1117/)

| Total | Passed | Failed | Flaky | Skipped |
|:-----:|:------:|:------:|:-----:|:-------:|
| 62 | 59 | 0 | 3 | 0 |

### Bundle Size: 🔺
Current: 78.88 MB | Main: 78.86 MB
Diff: +0.03 MB (0.04%)

⚠️ Bundle size increased. Please review.

<details>
<summary>ℹ️ CI Information</summary>

- Test recordings for failed tests are available in the full report.
- Bundle size is measured for the entire 'dist' directory.
- 📊 indicates links to detailed reports.
- 🔺 indicates increase, 🔽 decrease, and ✅ no change in bundle size.
</details>

### CODE REVIEW

File: src/containers/Tenant/Query/ExplainResult/ExplainResult.tsx [hunk 1]
```diff
@@ -153,43 +150,21 @@ export function ExplainResult({
...
-                                onCollapse={onCollapseResults}
-                                onExpand={onExpandResults}
-                                isCollapsed={isResultsCollapsed}
-                                initialDirection="bottom"
-                            />
-                        </div>
-                    </React.Fragment>
-                )}
-            </div>
+            <ResultControls<QueryExplainTab>
```
Comment (artemmufazalov) [inline src/containers/Tenant/Query/ExplainResult/ExplainResult.tsx:153]: It look like there is no much sense in splitting explain and execute results to different components now. Probably, it better merge them into one component, addded an issue for this: https://github.com/ydb-platform/ydb-embedded-ui/issues/1120

File: src/containers/Tenant/Query/ResultControls/ResultControls.tsx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            </div>
+            <div className={b('controls-left')}>
+                {clipboardText !== undefined && (
+                    <ClipboardButton
+                        text={clipboardText}
+                        view="flat-secondary"
+                        title="Copy results"
+                    />
+                )}
+                <EnableFullscreenButton disabled={isFullscreenDisabled} />
```
Comment (artemmufazalov) [inline src/containers/Tenant/Query/ResultControls/ResultControls.tsx:119]: Fullscreen button doesn't work until result is ready

File: tests/suites/tenant/queryEditor/consts.ts
Comment (artemmufazalov) [inline tests/suites/tenant/queryEditor/consts.ts:1]: Please, use full names, they are easier to read

File: tests/suites/tenant/queryEditor/consts.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+export const longRunningQuery = `
```
Comment (artemmufazalov) [inline tests/suites/tenant/queryEditor/consts.ts:1]: Please, add comment, that it's a long query that could cause "Memory limit exceeded" error, and that you should not run it on production DBs

File: src/containers/Tenant/Query/ResultControls/ResultControls.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                    </React.Fragment>
+                )}
+                {loading ? (
+                    <React.Fragment>
+                        <Label className={b('elapsed-label')}>
+                            {configuredNumeral(elapsedTime).format('00:00')}
+                        </Label>
+                        <Button onClick={onStopButtonClick} className={b('stop-button')}>
+                            <Icon data={StopFill} size={16} />
+                            {'Stop'}
```
Comment (artemmufazalov) [inline src/containers/Tenant/Query/ResultControls/ResultControls.tsx:106]: `i18n`

File: src/containers/Tenant/Query/ResultControls/ResultControls.tsx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                        </Button>
+                    </React.Fragment>
+                ) : null}
+            </div>
+            <div className={b('controls-left')}>
+                {clipboardText !== undefined && (
+                    <ClipboardButton
+                        text={clipboardText}
+                        view="flat-secondary"
+                        title="Copy results"
```
Comment (artemmufazalov) [inline src/containers/Tenant/Query/ResultControls/ResultControls.tsx:116]: `i18n`

**Review by artemmufazalov — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Query/ResultControls/ResultControls.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }, [loading]);
+
+    React.useEffect(() => {
+        intervalRef.current = window.setInterval(() => {
+            setElapsedTime(Math.floor((Date.now() - startTime) / SECOND_IN_MS));
+        }, SECOND_IN_MS);
+
+        return () => window.clearInterval(intervalRef.current);
+    }, [startTime]);
```
Comment (ValeraS) [inline src/containers/Tenant/Query/ResultControls/ResultControls.tsx:75]: Let's make component ElapsedTime:

```ts
import {duration} from '@gravity-ui/date-utils';

function ElapsedTime({className}: {className?: string}) {
    const [, reRender] = React.useState({});
    const [startTime] = React.useState(Date.now());
    const elapsedTime = Date.now() - startTime;

    React.useEffect(() => {
        const timerId = setInterval(() => {
            reRender({});
        }, SECOND_IN_MS);
        return () => {
            clearInterval(timerId);
        };
    }, []);

    // or formatDurationToShortTimeFormat(elapsedTime, 2)
    return <Label className={className}>{duration(elapsedTime).format('mm:ss')}</Label>;
}
```
and use it like this
```ts
{loading ? <ElapsedTime /> : nulll}
```

File: src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx [hunk 7]
```diff
@@ -194,23 +201,39 @@ function QueryEditor(props: QueryEditorProps) {
...
-            handleSendExecuteClick();
+        switch (lastUsedQueryAction) {
+            case QUERY_ACTIONS.explain:
+                handleGetExplainQueryClick();
+                break;
+            case QUERY_ACTIONS.execute:
+                handleSendExecuteClick();
+                break;
+            default:
+                break;
```
Comment (ValeraS) [inline src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx:236]: if `lastUsedQueryAction` is not defined we want to execute the query. And there are only two values, what is the use of the switch here?

File: src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx [hunk 8]
```diff
@@ -101,6 +103,7 @@ function QueryEditor(props: QueryEditorProps) {
     const [resultType, setResultType] = React.useState(RESULT_TYPES.EXECUTE);
     const [isResultLoaded, setIsResultLoaded] = React.useState(false);
     const [querySettings] = useQueryExecutionSettings();
+    const [currentQueryId, setCurrentQueryId] = React.useState<string | null>(null);
```
Comment (ValeraS) [inline src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx:106]: Maybe we can get current query id from `executeQueryResult` or `explainQueryResult`, something like this
```ts
let currentQueryId;
if (executeQueryResult.isLoading) {
    currentQueryId = executeQueryResult.originalArgs.queryId;
} else if (explainQueryResult.isLoading) {
    currentQueryId = explainQueryResult.originalArgs.queryId;
}
```

File: tests/suites/tenant/queryEditor/consts.ts
Comment (astandrik) [reply tests/suites/tenant/queryEditor/consts.ts:1]: Please make it more clear what do you mean by "full names"

Comment (astandrik) [reply tests/suites/tenant/queryEditor/consts.ts:1]: got it, you mean
constants

File: src/containers/Tenant/Query/ExecuteResult/ExecuteResult.tsx [hunk 9]
```diff
@@ -34,22 +30,24 @@ import './ExecuteResult.scss';
 
 const b = cn('ydb-query-execute-result');
 
-const resultOptionsIds = {
+const executeOptionsIds = {
```
Comment (ValeraS) [inline src/containers/Tenant/Query/ExecuteResult/ExecuteResult.tsx:33]: I don't see why these renamings are useful?

**Review by ValeraS — CHANGES_REQUESTED** [review]

**Review by Raubzeug — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Query/ResultControls/ResultControls.tsx [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                            </React.Fragment>
+                        )}
+                    </React.Fragment>
+                )}
+                {loading ? (
+                    <React.Fragment>
+                        <Label className={b('elapsed-label')}>
+                            {configuredNumeral(elapsedTime).format('00:00')}
+                        </Label>
+                        <Button onClick={onStopButtonClick} className={b('stop-button')}>
```
Comment (ValeraS) [inline src/containers/Tenant/Query/ResultControls/ResultControls.tsx:104]: Need to add a loading state while a cancel query running

File: src/containers/Tenant/Query/ResultControls/ResultControls.tsx (hunk 2 above)
Comment (astandrik) [reply src/containers/Tenant/Query/ResultControls/ResultControls.tsx]: got rid of ResultControls so now it works as it was before

File: src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (astandrik) [reply src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx:106]: used valeraS suggest with originalArgs

File: src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx [hunk 11]
```diff
@@ -194,18 +200,39 @@ function QueryEditor(props: QueryEditorProps) {
...
+        if (currentQueryId) {
+            sendCancelQuery({queryId: currentQueryId, database: tenantName});
+        }
+    }, [
+        executeQueryResult.isLoading,
+        executeQueryResult.originalArgs?.queryId,
+        explainQueryResult.originalArgs?.queryId,
+        sendCancelQuery,
+        tenantName,
+    ]);
```
[the remark this answers is not in the document]
Comment (astandrik) [reply src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx:234]: image
https://dev.to/kirubelkinfe/mastering-usecallback-in-react-with-typescript-4-different-use-cases-for-usecallback-3lga

[the remark this answers is not in the document]
Comment (astandrik) [reply src/containers/Tenant/Query/QueryEditor/QueryEditor.tsx]: moved currentQueryId outside

File: src/containers/Tenant/Query/ExecuteResult/ExecuteResult.tsx [hunk 12]
```diff
@@ -230,18 +237,36 @@ export function ExecuteResult({
...
+                                    />
+                                </React.Fragment>
+                            )}
                         </React.Fragment>
                     )}
+                    {loading ? (
+                        <React.Fragment>
+                            <div className={b('elapsed-time')}>
+                                <ElapsedTime />
+                            </div>
```
[the remark this answers is not in the document]
Comment (astandrik) [reply src/containers/Tenant/Query/ExecuteResult/ExecuteResult.tsx:268]: ElapsedTime is a self-contained label that counts time
it does not need any classes from outside

[the remark this answers is not in the document]
Comment (astandrik) [reply src/containers/Tenant/Query/ExecuteResult/ExecuteResult.tsx]: after discussion passed className as prop

**Review by Raubzeug — APPROVED** [review]

File: tests/suites/tenant/queryEditor/queryEditor.test.ts [hunk 13]
```diff
@@ -87,6 +88,22 @@ test.describe('Test Query Editor', async () => {
...
     });
 
+    test('Error is displayed for invalid query', async ({page}) => {
+        const queryEditor = new QueryEditor(page);
+
+        const invalidQuery = 'Select d';
+        await queryEditor.setQuery(invalidQuery);
+        await queryEditor.clickRunButton();
+
+        // Проверяем, что статус выполнения - "Failed"
```
Comment (artemmufazalov) [inline tests/suites/tenant/queryEditor/queryEditor.test.ts:98]: Please, write comments in English, we are open-source

**Review by artemmufazalov — CHANGES_REQUESTED** [review]

**Review by artemmufazalov — APPROVED** [review]

**Review by ValeraS — APPROVED** [review]

**Review by Raubzeug — APPROVED** [review]

**Pull request merged**